wagon-providers/wagon-http-shared/src/main/java/org/apache/maven/wagon/shared/http/AbstractHttpClientWagon.java

```diff
@@ -152,6 +154,49 @@ private WagonHttpEntity( final InputStream stream, final Resource resource, fina
             this.length = resource == null ? -1 : resource.getContentLength();
 
             this.wagon = wagon;
+
+            // if the autoset content flag is SET and the content type is blank
+            // then try to determine what the content type is and set it
+            if ( AUTOSET_CONTENT_TYPE && getContentType() == null )
+            {
+                setContentType( determineContentType() );
+            }
+        }
+
+        /**
+         * Best effort to determine the content type.
+         *
+         * if the content is coming from a file and the content type is determinable from the file extension
+         * or
+         * if the content is coming from a stream and the content type is determinable from the stream
+         *     (guessContentTypeFromStream will return null if the InputStream does not support mark())
+         * then determine and return the content type
+         * if the content type is not determinable then return "application/octet-stream"
+         *
+         * NOTE: this method is expected to always return a non-empty String
+         */
+        private String determineContentType()
+        {
+            try
+            {
+                String mimeType = this.source != null
+                            ? URLConnection.guessContentTypeFromName( this.source.getName() )
+                            : this.stream != null
+                                ? URLConnection.guessContentTypeFromStream( this.stream )
+                                : DEFAULT_CONTENT_TYPE;
```

Comment:
I consider this like to be redundant: RFC 7231, section 3.1.1.5:
>  If a Content-Type header field is not present, the recipient
   MAY either assume a media type of "application/octet-stream"
   ([RFC2046], Section 4.5.1) or examine the data to determine its type.

Reply:
I agree this is redundant, and I think most of the interest in this change is from people using client software that's bespoke, naive, or maybe completely ignorant of the RFC.

That said, I'd recommend doing it anyway. It's becoming apparent to me that there is a lot of client software out there that doesn't know/implement RFC 7231, section 3.1.1.5. It's perfectly _correct_ to not do this, but it's also a good way to keep getting asked the question by newbies (like I was). Unless you really want to keep educating people about it, I'd recommend just returning the default that the caller should assume.